### ISSUE TITLE
OmniFaces Gzip + PrimeFaces Data Exporter causes stack trace

### ISSUE THREAD

**Issue by melloware** [omnifaces/omnifaces#336]
OmniFaces: 2.5.1
PrimeFaces 6.0.12
Server: Wildfly 10.1

I have this in my PF page:

```xml
<h:commandLink id="lnkExportExcel" styleClass="ui-icon-export-excel" title="#{webmsg['action.export.excel.hint']}">
	<p:dataExporter type="xlsx" target="tblPerfStats" fileName="statistics-#{aboutController.getHost()}" />
</h:commandLink>
```
And I am getting this stack trace about the stream being closed:

```
11:10:46,910 WARNING [javax.enterprise.resource.webcontainer.jsf.lifecycle] (default task-29) java.io.IOException: Stream is already closed.: javax.faces.FacesException: java.io.IOException: Stream is already closed.
	at org.primefaces.component.export.DataExporter.processAction(DataExporter.java:131)
	at javax.faces.event.ActionEvent.processListener(ActionEvent.java:88)
	at javax.faces.component.UIComponentBase.broadcast(UIComponentBase.java:814)
	at javax.faces.component.UICommand.broadcast(UICommand.java:300)
	at javax.faces.component.UIViewRoot.broadcastEvents(UIViewRoot.java:790)
	at javax.faces.component.UIViewRoot.processApplication(UIViewRoot.java:1282)
	at com.sun.faces.lifecycle.InvokeApplicationPhase.execute(InvokeApplicationPhase.java:81)
	at com.sun.faces.lifecycle.Phase.doPhase(Phase.java:101)
	at com.sun.faces.lifecycle.LifecycleImpl.execute(LifecycleImpl.java:198)
	at javax.faces.webapp.FacesServlet.service(FacesServlet.java:658)
	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:85)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129)
	at org.omnifaces.filter.FacesExceptionFilter.doFilter(FacesExceptionFilter.java:93)
	at org.omnifaces.filter.HttpFilter.doFilter(HttpFilter.java:108)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at org.omnifaces.filter.GzipResponseFilter.doFilter(GzipResponseFilter.java:180)
	at org.omnifaces.filter.HttpFilter.doFilter(HttpFilter.java:108)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at org.primefaces.webapp.filter.FileUploadFilter.doFilter(FileUploadFilter.java:78)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at org.omnifaces.filter.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:123)
	at org.omnifaces.filter.HttpFilter.doFilter(HttpFilter.java:108)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at com.ksm.ee.web.servlet.SecureServletFilter.doFilter(SecureServletFilter.java:80)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at com.ksm.ee.web.servlet.MDCLogFilter.doFilter(MDCLogFilter.java:70)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at org.wildfly.extension.undertow.security.SecurityContextAssociationHandler.handleRequest(SecurityContextAssociationHandler.java:78)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:131)
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.server.handlers.DisableCacheHandler.handleRequest(DisableCacheHandler.java:33)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.security.handlers.AuthenticationConstraintHandler.handleRequest(AuthenticationConstraintHandler.java:53)
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
	at io.undertow.servlet.handlers.security.ServletSecurityConstraintHandler.handleRequest(ServletSecurityConstraintHandler.java:59)
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.security.handlers.NotificationReceiverHandler.handleRequest(NotificationReceiverHandler.java:50)
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.extension.undertow.security.jacc.JACCContextIdHandler.handleRequest(JACCContextIdHandler.java:61)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:292)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:138)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:135)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44)
	at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44)
	at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44)
	at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44)
	at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:272)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:104)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:202)
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:805)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: Stream is already closed.
	at org.omnifaces.servlet.GzipHttpServletResponse$GzipThresholdOutputStream.checkClosed(GzipHttpServletResponse.java:278)
	at org.omnifaces.servlet.GzipHttpServletResponse$GzipThresholdOutputStream.flush(GzipHttpServletResponse.java:215)
	at org.omnifaces.io.ResettableBufferedOutputStream.flush(ResettableBufferedOutputStream.java:92)
	at org.omnifaces.servlet.HttpServletResponseOutputWrapper$1.flush(HttpServletResponseOutputWrapper.java:104)
	at org.omnifaces.servlet.HttpServletResponseOutputWrapper.flushBuffer(HttpServletResponseOutputWrapper.java:147)
	at org.omnifaces.servlet.GzipHttpServletResponse.flushBuffer(GzipHttpServletResponse.java:116)
	at com.sun.faces.context.ExternalContextImpl.responseFlushBuffer(ExternalContextImpl.java:990)
	at javax.faces.context.ExternalContextWrapper.responseFlushBuffer(ExternalContextWrapper.java:940)
	at org.primefaces.component.export.ExcelExporter.writeExcelToResponse(ExcelExporter.java:258)
	at org.primefaces.component.export.ExcelExporter.export(ExcelExporter.java:86)
	at org.primefaces.component.export.DataExporter.processAction(DataExporter.java:120)
	... 72 more
```

**Comment by BalusC** [comment]
Reproduced. Doesn't happen with CSV and PDF.

**Comment by BalusC** [comment]
I debugged it, it appears that POI internally uses a ZipOutputStream which is already explicitly flushed and closed.

This means that PrimeFaces shouldn't need to flush the response at all in their [`ExcelExporter#writeExcelToResponse()`](https://github.com/primefaces/primefaces/blob/6_0/src/main/java/org/primefaces/component/export/ExcelExporter.java#L217). This is therefore a PrimeFaces bug. I suggest to report this issue to them.

**Comment by melloware** [comment]
Will report to PF thanks @BalusC  !

**Comment by melloware** [comment]
Follow up.  Reported it through my PRO account and PF has fixed it in 6.0.13 and 6.1.
